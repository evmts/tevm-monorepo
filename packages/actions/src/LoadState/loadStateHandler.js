import { InternalError } from '@tevm/errors'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateLoadStateParams } from './validateLoadStateParams.js'

/**
 * @internal
 * Creates a handler for loading a previously dumped state into the VM.
 *
 * @param {import("@tevm/node").TevmNode} client - The base client instance.
 * @param {object} [options] - Optional configuration.
 * @param {boolean} [options.throwOnFail] - Whether to throw an error when a failure occurs.
 * @returns {import('./LoadStateHandlerType.js').LoadStateHandler} - The handler function.
 *
 * @example
 * ```typescript
 * import { createClient } from 'tevm'
 * import { loadStateHandler } from 'tevm/actions'
 * import fs from 'fs'
 *
 * const client = createClient()
 * const loadState = loadStateHandler(client)
 *
 * const state = JSON.parse(fs.readFileSync('state.json'))
 * const result = await loadState({ state })
 * if (result.errors) {
 *   console.error('Failed to load state:', result.errors)
 * }
 * ```
 *
 * @see {@link LoadStateParams}
 * @see {@link LoadStateResult}
 * @see {@link TevmLoadStateError}
 */
export const loadStateHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params }) => {
		const errors = validateLoadStateParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}
		try {
			const vm = await client.getVm()
			if ('generateCanonicalGenesis' in vm.stateManager) {
				await vm.stateManager.generateCanonicalGenesis(params.state)
			} else {
				throw new Error(
					'Unsupported state manager. Must use a Tevm state manager from `@tevm/state` package. This may indicate a bug in tevm internal code.',
				)
			}
			return {}
		} catch (e) {
			return maybeThrowOnFail(throwOnFail, {
				errors: [new InternalError('UnexpectedError', { cause: e })],
			})
		}
	}
