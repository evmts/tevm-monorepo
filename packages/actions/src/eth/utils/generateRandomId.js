/**
 * @returns {import("@tevm/utils").Hex}
 */
export const generateRandomId = () => {
	return `0x${Array.from(crypto.getRandomValues(new Uint8Array(16)))
		.map((b) => b.toString(16).padStart(2, '0'))
		.join('')}`
}
