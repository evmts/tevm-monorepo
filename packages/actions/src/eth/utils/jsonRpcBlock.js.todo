import { bytesToHex } from "viem"

const ethjsTxToTevmTx = (tx) => {

}

/**
 * Transoforms an ethjs block to a tevm Block
 * @param {import("@ethereumjs/block").Block} block
 * @param {boolean} includeTransactions
 * @returns {Promise<import("@tevm/actions-types").Block>}
 */
const ethjsToTevmBlock = async (
  block,
  includeTransactions,
) => {
  const transactions = block.transactions.map((tx, txIndex) =>
    includeTransactions ? ethjsTxToTevmTx(tx, block, txIndex) : bytesToHex(tx.hash())
  )
  const withdrawalsAttr =
    block.header.withdrawalsRoot !== undefined
      ? {
        withdrawalsRoot: bytesToHex(block.header.withdrawalsRoot),
        withdrawals: block.toJSON().withdrawals,
      }
      : {}
  const td = await chain.getTd(block.hash(), block.header.number)
  return {
    coinbase: block.toJSON().header?.coinbase
    number: block.toJSON().header.number,
    hash: bytesToHex(block.hash()),
    parentHash: block.toJSON().header.parentHash,
    mixHash: block.toJSON().header.mixHash,
    nonce: block.toJSON().header.nonce,
    sha3Uncles: block.toJSON().header.uncleHash,
    logsBloom: block.toJSON().header.logsBloom,
    transactionsRoot: block.toJSON().header.transactionsTrie,
    stateRoot: block.toJSON().header.stateRoot,
    receiptsRoot: block.toJSON().header.receiptTrie,
    miner: block.toJSON().header.coinbase,
    difficulty: block.toJSON().header.difficulty,
    totalDifficulty: bigIntToHex(td),
    extraData: block.toJSON().header.extraData,
    size: intToHex(utf8ToBytes(JSON.stringify(block.toJSON())).byteLength),
    gasLimit: block.toJSON().header.gasLimit,
    gasUsed: block.toJSON().header.gasUsed,
    timestamp: block.toJSON().header.timestamp,
    transactions,
    uncles: block.uncleHeaders.map((uh) => bytesToHex(uh.hash())),
    baseFeePerGas: block.toJSON().header.baseFeePerGas,
    ...withdrawalsAttr,
    blobGasUsed: block.toJSON().header.blobGasUsed,
    excessBlobGas: block.toJSON().header.excessBlobGas,
    parentBeaconBlockRoot: block.toJSON().header.parentBeaconBlockRoot,
  }
}
