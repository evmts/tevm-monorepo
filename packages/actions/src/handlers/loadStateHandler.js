import { createError } from './createError.js'
import { Account, Address } from '@ethereumjs/util'
import { DefaultTevmStateManager, TevmStateManager } from '@tevm/state'
import { validateLoadStateParams } from '@tevm/zod'
import { fromRlp, hexToBytes, isHex } from 'viem'

/**
 * @param {TevmStateManager | DefaultTevmStateManager} stateManager
 * @returns {import('@tevm/actions-types').LoadStateHandler}
 */
export const loadStateHandler = (stateManager) => async (params) => {
	const errors = validateLoadStateParams(params)
	if (errors.length > 0) {
		return { errors }
	}

	const tevmState = params.state
	try {
		for (const [k, v] of Object.entries(tevmState)) {
			const { nonce, balance, storageRoot, codeHash, storage } = v
			const account = new Account(
				// replace with just the var
				nonce,
				balance,
				hexToBytes(storageRoot, { size: 32 }),
				hexToBytes(codeHash, { size: 32 }),
			)
			const address = Address.fromString(k)
			stateManager.putAccount(address, account)
			if (storage !== undefined) {
				for (const [storageKey, storageData] of Object.entries(storage)) {
					const key = hexToBytes(
						isHex(storageKey) ? storageKey : `0x${storageKey}`,
					)
					const encodedStorageData = fromRlp(
						isHex(storageData) ? storageData : `0x${storageData}`,
					)
					const data = hexToBytes(
						isHex(encodedStorageData)
							? encodedStorageData
							: `0x${encodedStorageData}`,
					)
					stateManager.putContractStorage(address, key, data)
				}
			}
		}
		return {}
	} catch (e) {
		return {
			errors: [
				createError(
					'UnexpectedError',
					typeof e === 'string'
						? e
						: e instanceof Error
						? e.message
						: 'unknown error',
				),
			],
		}
	}
}
