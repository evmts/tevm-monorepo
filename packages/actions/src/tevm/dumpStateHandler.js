import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import {
	ForkStateManager,
	NormalStateManager,
	ProxyStateManager,
} from '@tevm/state'

/**
 * @param {Pick<import("@tevm/base-client").BaseClient, 'getVm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').DumpStateHandler}
 */
export const dumpStateHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail } = {}) => {
		try {
			const vm = await client.getVm()
			if (
				vm.stateManager instanceof NormalStateManager ||
				vm.stateManager instanceof ProxyStateManager ||
				vm.stateManager instanceof ForkStateManager
			) {
				return { state: await vm.stateManager.dumpCanonicalGenesis() }
			} else {
				throw new Error(
					'Unsupported state manager. Must use a NormalStateManager, ProxyStateManager, or ForkStateManager. This indicates a bug in tevm internal code.',
				)
			}
		} catch (e) {
			return maybeThrowOnFail(throwOnFail ?? true, {
				state: {},
				errors: [
					createError(
						'UnexpectedError',
						typeof e === 'string'
							? e
							: e instanceof Error
							? e.message
							: 'unknown error',
					),
				],
			})
		}
	}
