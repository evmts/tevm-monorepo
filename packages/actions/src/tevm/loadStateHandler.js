import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import {
	ForkStateManager,
	NormalStateManager,
	ProxyStateManager,
} from '@tevm/state'
import { validateLoadStateParams } from '@tevm/zod'

/**
 * @param {Pick<import("@tevm/base-client").BaseClient, 'vm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').LoadStateHandler}
 */
export const loadStateHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params }) => {
		const errors = validateLoadStateParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}
		try {
			if (
				client.vm.stateManager instanceof NormalStateManager ||
				client.vm.stateManager instanceof ProxyStateManager ||
				client.vm.stateManager instanceof ForkStateManager
			) {
				await client.vm.stateManager.generateCanonicalGenesis(params.state)
			} else {
				throw new Error(
					'Unsupported state manager. Must use a NormalStateManager, ProxyStateManager, or ForkStateManager. This indicates a bug in tevm internal code.',
				)
			}
			return {}
		} catch (e) {
			return maybeThrowOnFail(throwOnFail, {
				errors: [
					createError(
						'UnexpectedError',
						typeof e === 'string'
							? e
							: e instanceof Error
							? e.message
							: 'unknown error',
					),
				],
			})
		}
	}
