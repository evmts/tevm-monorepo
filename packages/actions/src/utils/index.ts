export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
