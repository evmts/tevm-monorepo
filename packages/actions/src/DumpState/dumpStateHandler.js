import { InternalError } from '@tevm/errors'
import { bytesToHex } from '@tevm/utils'
import { getPendingClient } from '../internal/getPendingClient.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'

/**
 * Creates a handler for dumping the TEVM state.
 *
 * @param {import("@tevm/node").TevmNode} client - The TEVM client instance.
 * @param {object} [options] - Optional settings.
 * @param {boolean} [options.throwOnFail] - Whether to throw an error if the state dump fails.
 * @returns {import('../DumpState/DumpStateHandlerType.js').DumpStateHandler} - The state dump handler function.
 *
 * @example
 * ```typescript
 * import { createTevmNode } from 'tevm/node'
 * import { dumpStateHandler } from 'tevm/actions'
 *
 * const client = createTevmNode()
 *
 * const dumpState = dumpStateHandler(client)
 *
 * const { state, errors } = await dumpState()
 * if (errors) {
 *   console.error(errors)
 * } else {
 *   console.log(state)
 * }
 * ```
 */
export const dumpStateHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail, blockTag = 'latest' } = {}) => {
		try {
			const vm =
				blockTag === 'pending'
					? await getPendingClient(client).then((mineResult) => {
							if (mineResult.errors) {
								throw mineResult.errors[0]
							}
							return mineResult.pendingClient.getVm()
						})
					: await client.getVm()
			if ('dumpCanonicalGenesis' in vm.stateManager) {
				if (blockTag === 'latest' || blockTag === 'pending') {
					return { state: await vm.stateManager.dumpCanonicalGenesis() }
				}
				const block = await vm.blockchain.getBlockByTag(blockTag)
				if (await vm.stateManager.hasStateRoot(block.header.stateRoot)) {
					return { state: vm.stateManager._baseState.stateRoots.get(bytesToHex(block.header.stateRoot)) ?? {} }
				}
				client.logger.warn(`State root does not exist for block ${blockTag}. Returning empty state`)
				return { state: {} }
			}
			throw new InternalError(
				'Unsupported state manager. Must use a TEVM state manager from `@tevm/state` package. This may indicate a bug in TEVM internal code.',
			)
		} catch (e) {
			return maybeThrowOnFail(throwOnFail ?? true, {
				state: {},
				errors: [e instanceof InternalError ? e : new InternalError('Unexpected error', { cause: e })],
			})
		}
	}
