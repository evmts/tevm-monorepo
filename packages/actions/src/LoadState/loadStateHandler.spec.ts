import { createAddress } from '@tevm/address'
import { createTevmNode } from '@tevm/node'
import { createStateManager } from '@tevm/state'
import { bytesToHex, hexToBytes } from '@tevm/utils'
import { describe, expect, test } from 'vitest'
import { dumpStateHandler } from '../DumpState/dumpStateHandler.js'
import { loadStateHandler } from './loadStateHandler.js'

test('should load state into the state manager', async () => {
	const stateManager = createStateManager({})

	const address = createAddress('0x0420042004200420042004200420042004200420')

	let accountData = await stateManager.getAccount(address)

	// Expect state to be initially empty
	expect(accountData?.nonce).toBeUndefined()
	expect(accountData?.balance).toBeUndefined()

	const hashedStorageKey = '0x0c2d1b9c97b15f8a18e224fe94a8453f996465e14217e0939995ce76fbe01129'
	const storageValue = '0xa01000000000000000000000000000000000000000000000000000000000000000'

	//calls tevm state manager loadState method
	const state = {
		'0x0420042004200420042004200420042004200420': {
			nonce: '0x0',
			balance: '0x64',
			storageRoot: '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421' as const,

			codeHash: '0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470' as const,
			storage: {
				'0x0c2d1b9c97b15f8a18e224fe94a8453f996465e14217e0939995ce76fbe01129':
					'0xa01000000000000000000000000000000000000000000000000000000000000000',
			},
		},
	} as const

	const client = { getVm: () => ({ stateManager }) } as any

	await loadStateHandler(client)({ state, throwOnFail: false })

	accountData = await stateManager.getAccount(address)

	expect(accountData?.nonce).toEqual(0n)
	expect(accountData?.balance).toEqual(100n)

	const storedValue = await stateManager.getStorage(address, hexToBytes(hashedStorageKey))

	expect(bytesToHex(storedValue)).toEqualHex(storageValue)

	expect(await dumpStateHandler(client)()).toEqual({
		state: {
			'0x0420042004200420042004200420042004200420': {
				nonce: '0x0',
				balance: '0x64',
				storageRoot: '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421',
				codeHash: '0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470',
				storage: {
					'0x0c2d1b9c97b15f8a18e224fe94a8453f996465e14217e0939995ce76fbe01129':
						'0xa01000000000000000000000000000000000000000000000000000000000000000',
				},
			},
		},
	})
})

describe('loadStateHandler', () => {
	test('should return errors for invalid params', async () => {
		const client = createTevmNode()
		const handler = loadStateHandler(client)

		const result = await handler({
			state: 5 as any,
			throwOnFail: false,
		} as any)

		expect(result.errors).toBeDefined()
		expect(result.errors?.length).toBeGreaterThan(0)
		expect(result.errors?.[0]?.message).toMatchInlineSnapshot(`
			"Invalid state: Invalid input: expected record, received number

			Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
			Version: 1.1.0.next-73"
		`)
	})

	test('should throw error for unsupported state manager', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		// @ts-expect-error - Intentionally removing the method for testing
		delete vm.stateManager.generateCanonicalGenesis

		const handler = loadStateHandler({ getVm: () => Promise.resolve(vm) } as any)

		const result = await handler({ state: {}, throwOnFail: false })
		expect(result.errors?.[0]?.message).toMatchInlineSnapshot(`
"UnexpectedError

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Details: Unsupported state manager. Must use a Tevm state manager from \`@tevm/state\` package. This may indicate a bug in tevm internal code.
Version: 1.1.0.next-73"
`)
	})

	test('should handle error when generating genesis fails', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		vm.stateManager.generateCanonicalGenesis = () => {
			throw new Error('Genesis generation failed')
		}

		const handler = loadStateHandler({ getVm: () => Promise.resolve(vm) } as any)

		const result = await handler({
			state: {},
			throwOnFail: false,
		})

		expect(result.errors).toBeDefined()
		expect(result.errors?.length).toBe(1)
		expect(result.errors?.[0]?.message).toMatchInlineSnapshot(`
"UnexpectedError

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Details: Genesis generation failed
Version: 1.1.0.next-73"
`)
		expect(result.errors?.[0]?.cause?.message).toMatchInlineSnapshot(`"Genesis generation failed"`)
	})
})
