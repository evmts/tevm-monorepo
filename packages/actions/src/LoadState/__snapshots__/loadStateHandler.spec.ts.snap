// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadStateHandler > should return errors for invalid params 1`] = `
"state: state must be an object mapping addresses to account storage

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73"
`;

exports[`loadStateHandler should return errors for invalid params 1`] = `
"state: state must be an object mapping addresses to account storage

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73"
`;
